llvm/lib/IR/Instruction.cpp

```diff
@@ -698,6 +700,16 @@ bool Instruction::hasSameSpecialState(const Instruction *I2,
   assert(I1->getOpcode() == I2->getOpcode() &&
          "Can not compare special state of different instructions");
 
+  // MMRAs may change semantics of an operation, e.g. make a fence only
+  // affect a given address space.
+  //
+  // FIXME: Not sure if this stinks or not. Maybe we should just look at
+  // all callers and make them check MMRAs.
+  // OTOH, MMRAs can really alter semantics so this is technically correct
+  // (the best kind of correct).
+  if (MMRAMetadata(*this) != MMRAMetadata(*I2))
```

Comment:
This is something I'm unsure about. It doesn't feel like the right place to put this, but it kind of makes sense to put here given how MMRAs work?

Maybe it should check for compatibility instead? (@sameerds)

This is what prevents SimplifyCFG from merging common code. GVN does it though because it doesn't seem to use this function.

Reply:
This function must indeed not check metadata. Instead it will get resolved by metadata merging (implemented in Local.cpp).

Reply:
@nikic understood, so I need to teach each pass individually to check MMRAs and compare them? (= look at callers of this function and update them on a case-by-case basis)

Reply:
No, you should not need to do anything special. If instructions with different MMRAs get merged, then they will be updated correctly based on whatever merging you implemented.

Reply:
@nikic as @ssahasra said, we want to start by making pass treat MMRAs very conservatively, and avoid losing/combining them as much as possible. Over time we may loosen this up and allow for more merging once our toolchain uses MMRAs and we have more performance data to work with.

I implemented this restriction per-pass now (each pass checks MMRAs). I don't think we have that many passes that can merge memory/synchronizing instructions so I hope I covered most of them, is there any that are missing?
(note: GVN is still missing, I have to figure out how to handle that one)

Reply:
I think if you want to aggressively not merge MMRAs, then you shouldn't be modelling them using metadata.

Reply:
It's one of the major questions I had in the RFC, but it unfortunately didn't gather much attention on Discourse (see "Next Steps"): https://discourse.llvm.org/t/rfc-mmras-memory-model-relaxation-annotations/76361

I can see pros and cons for MD:
- (+) Metadata is less intrusive and models the optional nature of MMRAs well (= dropping them never affects correctness, unlike, say, dropping syncscope)
- (+) Even if we want to avoid combining MMRAs, it's still a well-defined process. We just want to avoid doing it aggressively  (in this case, we want to avoid doing it as part of the IR canonicalization passes such as SimplifyCFG, IC, ...)
- (-) Metadata isn't an intuitive way to model a property that can alter semantics, but there are precedents for that (e.g. !amdgpu.uniform & similar MD used to guide codegen) 
- (-) Metadata means pass writers are more likely to just drop it (or forget it), so it wouldn't surprise me if we get issues every now and then due to a commit dropping MMRAs in X or Y pass (but I don't think it'll be a very common occurrence either)

My proposal would be to start MMRAs as metadata because it's non-intrusive. After all, this is something new, and only time will tell how much use it gets. We have AMDGPU use cases that were tested internally, but we don't know if other backends will make use of them. 
We can reevaluate if needed later.

Reply:
The way you are approaching this, you can't *really* drop MMRAs. Like, yes, it's semantically correct to do, but it sounds like every time it would happen in a generic transform, you would actually treat that like a bug and slap an extra MMRA check in that transform. You are actively fighting the metadata model at every step here.

I'm not sure to what degree all these checks you have introduced are *actually* necessary, as opposed to you being overly conservative. For example, is `if (x) { load p, !mmra1 } else { load p, !mmra2 }` really something that you expect to happen a lot in practice? But if these are necessary, then I don't think this is the right mechanism.

Reply:
Most of these checks are necessary, for instance one of the use case we tested basically has (in pseudocode):
```
if (A) 
  fence release, !mmra !0
else if (B)
  fence release, !mmra !1
else 
  fence release, !mmra !2
```
And SimplifyCFG would just remove all CF and leave `fence release`, undoing all the work. Such cases would arise in OpenCL code whenever one or two address spaces are being fenced instead of all of them.

I'm really hesitant about making this an instruction operand. I really don't mind doing the work if it makes a better end result, but I'm not yet convinced it's strictly better. 

This is unfortunately right in the middle between MD and operand so I guess which one should be used depends on the point of view. IMO it's better to make it metadata because this has much more in common with other MD nodes than with something like syncscope (which can never be dropped/merged/etc.)

Also we wouldn't treat every instance of dropped metadata like a bug, it's more nuanced. Dropping MMRAs doesn't affect correctness, it just has a performance cost of X, but if an optimization being disabled has a performance cost of Y and Y > X, then the optimization must take priority and be allowed to run and drop MMRAs anyway.

If the amount of pass changes is going to be a blocker I'm open to taking a less conservative approach and just leave the Transform changes that I can prove actively hurt MMRAs (so if I have real-world evidence that that pass caused problems). IIRC that'd only leave SimplifyCFG but I'd need to run more tests. I can then make more changes over time if needed. Would that be better?

Reply:
> If the amount of pass changes is going to be a blocker I'm open to taking a less conservative approach and just leave the Transform changes that I can prove actively hurt MMRAs (so if I have real-world evidence that that pass caused problems). IIRC that'd only leave SimplifyCFG but I'd need to run more tests. I can then make more changes over time if needed. Would that be better?

Yes, that would be better. I think your fence example for SimplifyCFG is compelling, but I'm still doubtful about other changes you've made (esp. when it comes to CSE-like transforms).

Reply:
> If the amount of pass changes is going to be a blocker I'm open to taking a less conservative approach and just leave the Transform changes that I can prove actively hurt MMRAs (so if I have real-world evidence that that pass caused problems). IIRC that'd only leave SimplifyCFG but I'd need to run more tests. I can then make more changes over time if needed. Would that be better?

I agree.